rs/index/src/ivf/writer.rs

```diff
@@ -167,12 +161,51 @@ impl<Q: Quantizer> IvfWriter<Q> {
     }
 
     fn write_posting_lists_and_metadata(&self, ivf_builder: &mut IvfBuilder) -> Result<usize> {
-        let path = format!("{}/posting_lists", self.base_directory);
-        let mut file = File::create(path)?;
-        let mut writer = BufWriter::new(&mut file);
+        let metadata_path = format!("{}/posting_list_metadata", self.base_directory);
+        let mut metadata_file = File::create(metadata_path)?;
+        let mut metadata_writer = BufWriter::new(&mut metadata_file);
+
+        let posting_list_path = format!("{}/posting_lists", self.base_directory);
+        let mut posting_list_file = File::create(posting_list_path)?;
+        let mut posting_list_writer = BufWriter::new(&mut posting_list_file);
+
+        let mut metadata_bytes_written = 0;
+        let mut posting_list_bytes_written = 0;
+
+        let num_posting_lists = ivf_builder.posting_lists().len();
+        // First write the total number of posting lists
+        metadata_bytes_written +=
+            wrap_write(&mut metadata_writer, &num_posting_lists.to_le_bytes())?;
+        for i in 0..num_posting_lists {
+            let posting_list = ivf_builder
+                .posting_lists()
+                .get(i as u32)?
+                .iter()
+                .collect::<Vec<_>>();
+            // Write the length of the posting list
+            metadata_bytes_written +=
+                wrap_write(&mut metadata_writer, &posting_list.len().to_le_bytes())?;
+            // Write the offset to the current posting list
+            metadata_bytes_written += wrap_write(
+                &mut metadata_writer,
+                &((posting_list_bytes_written as u64).to_le_bytes()),
+            )?;
+            // Now write the posting list itself
+            let mut encoder = C::new_encoder(/*universe*/ None, posting_list.len());
+            encoder.encode(&posting_list)?;
+            posting_list_bytes_written += encoder.write(&mut posting_list_writer)?;
```

Comment:
you're going to want to encode the list first, so that you know the length of the encoded buffer before writing the length out. Also it's probably a good idea to write the encoded buffer length instead of number of elements.